File: Resmat1/iago/trabalho.cpp
#include <iostream>
#include <cmath>
#include <fstream>
#include <string>

using namespace std;

void salvarArquivo (int figuras, string nome, int nc, double totalArea, double totalPerimetro, double totalCentroX, double totalCentroY, double totalInerciaX, double totalInerciaY, 
					double totalInerciaXY, double totalInceriaPolar, double momentoInerciaMin, double momentoInceriaMax, double anguloTeta1, double anguloTeta2, double raioGiroMin, double raioGiroMax, string unidade) 
{
	
	
	string snome;
	snome = nome + ".OUT"; //concatenar nome para salvar arquivo
	ofstream escreverArq(snome.c_str(),ios::binary);
	if (escreverArq.is_open() ){
		escreverArq << fixed;
		escreverArq.precision(6);
		//dados impressos
		escreverArq << "      **** PROPRIEDADES GEOMETRICAS DAS FIGURAS PLANAS ****"<<endl;
		escreverArq << "                    RESULTADOS OBTIDOS:"<<endl<<endl;
		escreverArq << " NUMERO DE FIGURAS NO CONTORNO:    		"<<figuras<<endl;
		escreverArq << " NUMERO DE VERTICES NO CONTORNO:    	"<<nc<<endl<<endl;
		escreverArq << " AREA DA FIGURA:                     	"<<totalArea<<unidade<<"2"<<endl; 
		escreverArq << " PERIMETRO DA FIGURA:                   "<<totalPerimetro<<unidade<<""<<endl;  
		escreverArq <<endl;
		escreverArq <<"	COORDENADA X DO C. G.:                  "<<totalCentroX<<unidade<<""<<endl;  
		escreverArq <<"	COORDENADA Y DO C. G.:                  "<<totalCentroY<<unidade<<""<<endl;  
		escreverArq <<endl;
		escreverArq <<"	MOMENTO DE INERCIA, inerciaX:           "<<totalInerciaX<<unidade<<"4"<<endl; 
		escreverArq <<"	MOMENTO DE INERCIA, inerciaY:           "<<totalInerciaY<<unidade<<"4 "<<endl;
		escreverArq <<"	PRODUTO DE INERCIA, inerciaXY:          "<<totalInerciaXY<<unidade<<"4 "<<endl;
		escreverArq <<"	MOMENTO POLAR DE INERCIA, Ip:			"<<totalInceriaPolar<<unidade<<"4"<<endl;
		escreverArq <<endl;
		escreverArq <<"	MOMENTO DE INERCIA MINIMO, Imn:         "<<momentoInerciaMin<<unidade<<"4 "<<endl;
		escreverArq <<"	MOMENTO DE INERCIA MAXIMO, Imx:         "<<momentoInceriaMax<<unidade<<"4 "<<endl;
		escreverArq <<endl;
		escreverArq <<"	ANG. INCL. EIXO PRINC. - Teta1:         "<<anguloTeta1<<"°   "<<endl;
		escreverArq <<"	ANG. INCL. EIXO PRINC. - Teta2:         "<<anguloTeta2<<"°   "<<endl;
		escreverArq <<endl;
		escreverArq <<"	RAIO DE GIRACAO, Rmin.:                 "<<raioGiroMin<<unidade<<""<<endl;
		escreverArq <<"	RAIO DE GIRACAO, RmAx:			    	"<<raioGiroMax<<unidade<<""<<endl;
	 }
	 escreverArq.close();
}
int main ()
{
	//Entradas de dados
	string nome, lnome, unidade;
	cout << "Nome do arquivo: ";
	cin >> nome;
	
	lnome = nome+".DAT"; //concatenar nome do arquivo com extensao
	ifstream lerArq(lnome.c_str(),ios::binary); //nome passado do arquivo
	
	if(!lerArq.is_open( ))  //verifica se o programa conseguiu abrir o arquivo de nome informado pelo usuario *
    {
		cout<<"Não foi possível abrir arquivo!\n";
        lerArq.clear( ); //reseta o objeto lerArq
    }
	
	//lendo os elementos no arquivo
	int nc=0, figuras;
	lerArq >> figuras; //quantidade de figuras
	
	double nvc[figuras]; //coordenadas e numero de vertices de cada contorno
	int i,j,andar, identificador, inicioFigura;
	
	for (i=0; i<figuras; i++)
	{
		lerArq >> nvc[i];
		nc = nc + nvc[i];
	}
	double x[nc], y[nc];
	for (i=0; i<nc; i++)
	{
		lerArq >> x[i];
		lerArq >> y[i]; 
	}
	lerArq >> unidade;
	lerArq.close();//fecha arquivo
	
	/*------------Area e Perimetro-------*/
	double area[figuras], totalArea, perimetro[figuras], totalPerimetro;
	/*----------Centro de Gravidade-----------*/
	double centroGravidadeX[figuras], totalCentroX, centroGravidadeY[figuras], totalCentroY;
	
	
	//Inicializa a variavel em 0 para ser feito o somatório e não pegar lixo
	totalArea = 0; //total da area
	totalPerimetro = 0; //total do perimetro
	totalCentroX = 0; //total centro de gravidade em x
	totalCentroY = 0; //total centro de gravidade me y

	andar=0; //variavel andar utilizada para percorrer o vetor das coordenadas x e y
	identificador=0; //variavel utilizada para identificar quando a figura fecha
	inicioFigura=0; //variavel que é responsavel por aderir a posição do começo da figura para poder fechar ela
	
	for (i=0; i<figuras; i++) //percorre a quantidade de figuras digitadas 
	{
		//inicializamos em 0 para somatorio
		area[i] = 0; 
		perimetro[i] = 0;
		centroGravidadeX[i] = 0;
		centroGravidadeY[i] = 0;
		for (j=0; j<nvc[i]; j++) //percorremos a quantidade de aresta de cada figura
		{
			if (identificador==(nvc[i]-1)) //if que identifica se chegou no final da figura para fechar com o inicio
			{
				//inicioFigura identifica a coordenada inicial para fechar a figura
				//equações para calculos
				perimetro[i] = perimetro[i] + sqrt(pow((x[inicioFigura]-x[andar]),2) + pow((y[inicioFigura]-y[andar]),2)); 
				area[i] = area[i] + (x[andar]*y[inicioFigura]-x[inicioFigura]*y[andar]);
				centroGravidadeX[i] = centroGravidadeX[i] + ((x[andar]+x[inicioFigura])*(x[andar]*y[inicioFigura]-x[inicioFigura]*y[andar]));
				centroGravidadeY[i] = centroGravidadeY[i] + ((y[andar]+y[inicioFigura])*(x[andar]*y[inicioFigura]-x[inicioFigura]*y[andar]));
			}else
			{
				//andar e andar+1 percorrer as coordenadas
				perimetro[i] = perimetro[i] + sqrt(pow((x[andar+1]-x[andar]),2) + pow((y[andar+1]-y[andar]),2));
				area[i] = area[i] + (x[andar]*y[andar+1]-x[andar+1]*y[andar]);
				centroGravidadeX[i] = centroGravidadeX[i] + ((x[andar]+x[andar+1])*(x[andar]*y[andar+1]-x[andar+1]*y[andar]));
				centroGravidadeY[i] = centroGravidadeY[i] + ((y[andar]+y[andar+1])*(x[andar]*y[andar+1]-x[andar+1]*y[andar]));
			}
			andar++; //incrementa para percorrer coordenada
			identificador++; //incrementa para chegar no final da figura
		}
		identificador = 0; //zera para poder começar outra figura
		inicioFigura = nvc[i]+inicioFigura; //pega o ponto inicial da proxima figura
		
		//somatorio das equações
		totalPerimetro = totalPerimetro + perimetro[i];
		totalArea = totalArea + area[i];
		totalCentroX = totalCentroX + centroGravidadeX[i];
		totalCentroY = totalCentroY + centroGravidadeY[i];
	}
	totalArea = 0.5*totalArea; //calculo area
	totalCentroX = totalCentroX/(6*totalArea); //calculo centro de gravidade x 
	totalCentroY = totalCentroY/(6*totalArea); // calculo centro de gravidade y
	
	/*-----------Momento de Inércia--------------*/
	//Subtração das coordenas dos seus centro de gravidades para o calculo da Inercia
	for (i=0; i<nc; i++)
	{
		x[i] = x[i] - totalCentroX;
		y[i] = y[i] - totalCentroY;
	}
	
	//declaração de variaveis
	double inerciaX[figuras], totalInerciaX, inerciaY[figuras], totalInerciaY, inerciaXY[figuras], totalInerciaXY, totalInceriaPolar, a[nc], somaIX, somaIY, somaIXY, aCalc;
	//inicialização em 0 para somatorio
	totalInerciaX = 0;
	totalInerciaY = 0;
	totalInerciaXY = 0;
	somaIX = 0;
	somaIY = 0;
	somaIXY = 0;
	aCalc = 0;
	andar=0; //variavel andar utilizada para percorrer o vetor das coordenadas x e y
	identificador=0; //variavel utilizada para identificar quando a figura fecha
	inicioFigura=0; //variavel que é responsavel por aderir a posição do começo da figura para poder fechar ela
	
	for (i=0; i<figuras; i++) //percorre a quantidade de figuras digitadas 
	{
		for (j=0; j<nvc[i]; j++) //percorremos a quantidade de aresta de cada figura
		{
			if (identificador==(nvc[i]-1)) //if que identifica se chegou no final da figura para fechar com o inicio
			{
				//inicioFigura identifica a coordenada inicial para fechar a figura
				//equações para calculos
				aCalc = ((x[andar]*y[inicioFigura])-(x[inicioFigura]*y[andar]));
				a[andar] = aCalc;
				somaIX = somaIX + a[andar]*(pow(y[andar],2)+y[andar]*y[inicioFigura]+pow(y[inicioFigura],2));
				somaIY = somaIY + a[andar]*(pow(x[andar],2)+x[andar]*x[inicioFigura]+pow(x[inicioFigura],2));
				somaIXY = somaIXY + a[andar]*(x[andar]*y[inicioFigura]+2*(x[andar]*y[andar])+2*(x[inicioFigura]*y[inicioFigura])+x[inicioFigura]*y[andar]);
			}else
			{
				//andar e andar+1 percorrer as coordenadas
				aCalc = ((x[andar]*y[andar+1])-(x[andar+1]*y[andar]));
				a[andar] = aCalc;
				somaIX = somaIX + a[andar]*(pow(y[andar],2)+y[andar]*y[andar+1]+pow(y[andar+1],2));
				somaIY = somaIY + a[andar]*(pow(x[andar],2)+x[andar]*x[andar+1]+pow(x[andar+1],2));
				somaIXY = somaIXY + a[andar]*(x[andar]*y[andar+1]+2*(x[andar]*y[andar])+2*(x[andar+1]*y[andar+1])+x[andar+1]*y[andar]);
			}
			andar++; //incrementa para percorrer coordenada
			identificador++; //incrementa para chegar no final da figura
		}
		identificador = 0; //zera para poder começar outra figura
		inicioFigura = nvc[i]+inicioFigura; //pega o ponto inicial da proxima figura
		
		//somatorio das equações
		inerciaX[i] = somaIX;
		inerciaY[i] = somaIY;
		inerciaXY[i] = somaIXY;
		somaIX=0;
		somaIY=0;
		somaIXY=0;
	}
	
	for (i=0; i<figuras; i++) //percorre a quantidade de figuras digitadas 
	{
		//somatorio
		totalInerciaX = totalInerciaX + inerciaX[i];
		totalInerciaY = totalInerciaY + inerciaY[i];
		totalInerciaXY = totalInerciaXY +inerciaXY[i];
	}
	//calculo final da equação
	totalInerciaX = totalInerciaX/12;
	totalInerciaY = totalInerciaY/12;
	totalInerciaXY = totalInerciaXY/24;
	totalInceriaPolar = totalInerciaX + totalInerciaY;

	//Abaixo só o uso de equações
	/*------Momento de Inercia Minimo e Maximo------*/
	double momentoInceriaMax, momentoInerciaMin;
	
	momentoInceriaMax = (totalInerciaX+totalInerciaY)/2 + sqrt(pow(((totalInerciaX-totalInerciaY)/2),2)+pow(totalInerciaXY,2));
	momentoInerciaMin = (totalInerciaX+totalInerciaY)/2 - sqrt(pow(((totalInerciaX-totalInerciaY)/2),2)+pow(totalInerciaXY,2));
	
	
	/*-------TetaP--------*/
	double anguloTeta1, anguloTeta2;
	
	anguloTeta1 = (atan(-totalInerciaXY/((totalInerciaX-totalInerciaY)/2)))/2;
	anguloTeta1 = anguloTeta1*180/M_PI;
	anguloTeta2 = anguloTeta1 + 90;
	
	
	/*--------Raio de giração------------*/
	double raioGiroMin, raioGiroMax;
	
	raioGiroMin = sqrt(momentoInerciaMin/totalArea);
	raioGiroMax = sqrt(momentoInceriaMax/totalArea);
	
	//chama funcao que salva no arquivo
	salvarArquivo(figuras, nome, nc, totalArea, totalPerimetro, totalCentroX, totalCentroY, totalInerciaX, totalInerciaY, totalInerciaXY, 
				totalInceriaPolar, momentoInerciaMin, momentoInceriaMax, anguloTeta1, anguloTeta2, raioGiroMin, raioGiroMax, unidade);
}

File: Resmat1/Fernando/trabalho.py
from math import* #importa funções matemáticas
import math
#Função para pegar os valores do arquivo
def criar(): #funcao para ler os dados dos arquivos
	print ("Digite o nome do arquivo: ")
	arq=input() #digitar nome do arquivo 
	arqentrada = arq + '.dat' #junção de string para ler o nome do arquivo
	f = open(arqentrada, 'r') #abrir arquivo para leitura
	#ler as entradas do arquivo
	figuras = int(f.readline())
	pontos = 0
	arestas=[]
	for i in range(figuras):
		aux = int(f.readline())
		pontos = pontos + aux
		arestas.append(aux)
	px=[] #vetor coordenada x
	py=[] #vetor coordenada y
	for i in range(pontos):
		x, y = map(float, f.readline().split())
		px.append(x)
		py.append(y)
	unidade = f.readline()
	return figuras,pontos,arestas,px,py,arq,unidade

def principal():
	figuras,pontos,arestas,x,y,arqsaida,unidade=criar()
	i=0 #laço pra percorrer as figuras 
	j=0 #laço pra percorrer as arestas de cada figura
	k=0 #variavel para andar nos pontos
	aux=0 #aux identifica quando a figura estiver na ultimo ponto para poder fechar com o primeiro ponto
	prox=0 #prox armazena a posição inicial da figura
	area=[]
	tarea=0.0
	perimetro=[]
	tperimetro=0.0
	area.append(0.0)
	perimetro.append(0.0)
	somap=0
	somaa=0
	
	#laço para fazer as equações
	#perimetro e area
	for i in range(figuras): #laço percorre as figuras 
		for j in range(arestas[i]): #laço percorre 
			if(aux==(arestas[i]-1)): #verifica final do arquivo
				somap+=sqrt(((x[prox]-x[k])**2)+((y[prox]-y[k])**2))
				somaa+=x[k]*y[prox]-x[prox]*y[k]
			else: 
				somap+=sqrt(((x[k+1]-x[k])**2)+((y[k+1]-y[k])**2))
				somaa+=x[k]*y[k+1]-x[k+1]*y[k]
			k+=1 #k percorre as figuras
			aux+=1 #aux anda até final da figura
		perimetro.append(somap) #armazena perimetro de cada figura no vetor
		area.append(somaa) #armazena area de cada figura no vetor
		aux=0 #zera aux pra proxima figura
		prox=arestas[i]+prox #prox é atualizado pra o novo inicio
		somaa=0 #zera variaveis
		somap=0

	for i in range(len(area)):#somar area,somar perimetro
		tarea+=area[i]
		tperimetro+=perimetro[i]
	tarea*=0.5
		
	#cacluco centro de gravidade x,y
	cgx=[]
	tcgx=0.0
	cgy=[]
	tcgy=0.0
	k=0
	aux=0
	prox=0
	cgx.append(0.0)
	cgy.append(0.0)
	somax=0
	somay=0
	for i in range(figuras):
		for j in range(arestas[i]):
			if(aux==(arestas[i]-1)):
				somax+=((x[k]+x[prox])*(x[k]*y[prox]-x[prox]*y[k]))
				somay+=((y[k]+y[prox])*(x[k]*y[prox]-x[prox]*y[k]))
			else:
				somax+=((x[k]+x[k+1])*(x[k]*y[k+1]-x[k+1]*y[k]))
				somay+=((y[k]+y[k+1])*(x[k]*y[k+1]-x[k+1]*y[k]))
			k+=1
			aux+=1
		cgx.append(somax)
		cgy.append(somay)
		aux=0
		prox+=arestas[i]
		somax=0
		somay=0

	for i in range(len(cgx)):
		tcgx+=cgx[i]
		tcgy+=cgy[i]

	tcgx/=(6*tarea)
	tcgy/=(6*tarea)

	#momento inercia
	for i in range(pontos): #atualiza os valores do x e y para calculo do momento de inercia
		x[i] = x[i] - tcgx
		y[i] = y[i] - tcgy
	Ix=[]
	tIx=0.0
	Iy=[]
	tIy=0.0
	Ixy=[]
	tIxy=0.0
	a=[]
	k=0
	aux=0
	prox=0
	Ix.append(0.0)
	Iy.append(0.0)
	Ixy.append(0.0)
	somaix=0
	somaiy=0
	somaixy=0
	ax=0
	for i in range(figuras):
		for j in range(arestas[i]):
			if(aux==(arestas[i]-1)):
				ax=((x[k]*y[prox])-(x[prox]*y[k]))
				a.append(ax)
				somaix+=a[k]*(pow(y[k],2)+y[k]*y[prox]+pow(y[prox],2))
				somaiy+=a[k]*(pow(x[k],2)+x[k]*x[prox]+pow(x[prox],2))
				somaixy+=a[k]*(x[prox]*y[k]+x[k]*y[prox]+2*x[k]*y[k]+2*x[prox]*y[prox])
			else:
				ax=((x[k]*y[k+1])-(x[k+1]*y[k]))
				a.append(ax)
				somaix+=a[k]*(pow(y[k],2)+y[k]*y[k+1]+pow(y[k+1],2))
				somaiy+=a[k]*(pow(x[k],2)+x[k]*x[k+1]+pow(x[k+1],2))
				somaixy+=a[k]*(x[k+1]*y[k]+x[k]*y[k+1]+2*x[k]*y[k]+2*x[k+1]*y[k+1])
			k+=1
			aux+=1
		aux=0
		prox=arestas[i]+prox
		Ix.append(somaix)
		Iy.append(somaiy)
		Ixy.append(somaixy)
		somaix=0
		somaiy=0
		somaixy=0
			
	for i in range(len(Ix)):
		tIx+=Ix[i]
		tIy+=Iy[i]
		tIxy+=Ixy[i]

	tIx = tIx/12
	tIy = tIy/12
	tIxy = tIxy/24
	poI = tIx + tIy


	#Momento de inercia maximo e minimo
	Imax = (tIx+tIy)/2 + sqrt(((tIx-tIy)/2)**2+(tIxy**2))

	Imin = (tIx+tIy)/2 - sqrt(((tIx-tIy)/2)**2+tIxy**2)

	#angulo de inclinacao x e y
	tetap1 = (math.atan(-tIxy/((tIx-tIy)/2)))/2 #math.atan arcotagente

	tetap1 = tetap1*180/math.pi #numero de pi

	tetap2 = tetap1 + 90
	
	#raio de giracao min e max
	Kx = sqrt(Imax/tarea)
	Ky = sqrt(Imin/tarea)


	#salvando o arquivo
	arqsaida = arqsaida + '.out'
	arquivo = open(arqsaida,'w')#abrir para escrever no arquivo
	arquivo.write("**** PROPRIEDADES GEOMÉTRICAS DAS FIGURAS PLANAS **** \n"
	"				RESULTADOS OBTIDOS: \n \n "
	"NUMERO DE CONTORNOS NA FIGURA:  "+str(figuras)+"\n \n"
	"NUMERO DE VERTICES NO CONTORNO: "+str(pontos)+"\n \n"
	"A area da sua figura e: 				{:.6f}".format(tarea)+"		"+str(unidade)+"² \n "
	"O perimetro de sua figura e: 			{:.6f}".format(tperimetro)+"		cm \n"
	"Coordenada do CG no eixo x: 			{:.6f}".format(tcgx)+"		"+str(unidade)+" \n"
	"Coordenada do CG no eixo y: 			{:.6f}".format(tcgy)+"		"+str(unidade)+" \n \n"
	"Momento de inercia em x: 				{:.6f}".format(tIx)+"		"+str(unidade)+"4  \n"
	"Momento de inercia em y: 				{:.6f}".format(tIy)+"		"+str(unidade)+"4  \n"
	"Produto de inercia em xy: 				{:.6f}".format(tIxy)+"		"+str(unidade)+"4  \n"
	"Momento polar de inercia, Ip: 			{:.6f}".format(poI)+"		"+str(unidade)+"4  \n \n"
	"Momento de inércia mínimo: 				{:.6f}".format(Imin)+"		"+str(unidade)+"4  \n"
	"Momento de inércia máximo: 				{:.6f}".format(Imax)+"		"+str(unidade)+"4  \n"
	"O Tetap1 de sua figura é: 				{:.9f}".format(tetap1)+"º \n"
	"O Tetap2 de sua figura é: 				{:.9f}".format(tetap2)+"º \n \n"
	"O Raio de giração, Rmin: 				{:.6f}".format(Ky)+"	"+str(unidade)+" \n"
	"O Raio de giração, Rmax: 				{:.6f}".format(Kx)+"	"+str(unidade)+" \n")
	arquivo.close()
	
	print ("Arquivo salvo.")
principal()




	
		
